File: JDK11/opentraceability/src/main/java/opentraceability/utility/UOMS.java
package opentraceability.utility;

import opentraceability.OTLogger;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class UOMS {
    public UOMS() throws Exception {
        load();
    }

    public static Map<String, UOM> uomsAbbrevDict = new HashMap<>();
    public static Map<String, UOM> uomsUNCodeDict = new HashMap<>();

    public static void load() throws Exception {
        uomsAbbrevDict = new HashMap<>();
        uomsUNCodeDict = new HashMap<>();

        // Load the subscriptions JSON
        EmbeddedResourceLoader loader = new EmbeddedResourceLoader();
        String jsonText = loader.readString(UOMS.class, "/uoms.json");
        JSONArray jarr = new JSONArray(jsonText);

        for (int i = 0; i < jarr.length(); i++) {
            JSONObject juom = jarr.getJSONObject(i);
            UOM uom = new UOM(juom);
            String lowerCaseAbbreviation = uom.Abbreviation.toLowerCase();
            String upperCaseUNCode = uom.UNCode.toUpperCase();

            if (!uomsAbbrevDict.containsKey(lowerCaseAbbreviation)) {
                uomsAbbrevDict.put(lowerCaseAbbreviation, uom);
            } else {
                //System.out.println("Duplicate Unit abbreviation detected: " + uom.Abbreviation);
            }

            if (!uomsUNCodeDict.containsKey(upperCaseUNCode)) {
                uomsUNCodeDict.put(upperCaseUNCode, uom);
            } else {
                //System.out.println("Duplicate Unit UNCode detected: " + uom.UNCode);
            }
        }
    }

    public static UOM getBase(UOM uom) throws Exception {
        return getBase(uom.UnitDimension);
    }

    public static UOM getBase(String dimension) throws Exception {
        for (UOM entry : uomsAbbrevDict.values()) {
            if (entry.UnitDimension.equals(dimension) && entry.isBase()) {
                return entry;
            }
        }
        throw new Exception("Failed to get base for dimension = " + dimension);
    }

    public static UOM getUOMFromName(String name) {
        UOM uom = null;
        String formattedName = name.toLowerCase();
        if (formattedName.equals("count")) {
            formattedName = "ea";
        }
        if (formattedName.equals("pound") || formattedName.equals("pounds") || formattedName.equals("ib")) {
            formattedName = "lb";
        }
        if (formattedName.charAt(formattedName.length() - 1) == '.') {
            formattedName = formattedName.substring(0, formattedName.length() - 1);
        }
        if (uomsAbbrevDict.containsKey(formattedName)) {
            uom = uomsAbbrevDict.get(formattedName);
        } else {
            for (UOM entry : uomsAbbrevDict.values()) {
                if (entry.Name.toLowerCase().equals(formattedName)) {
                    uom = entry;
                    break;
                }
            }
            if (uom == null) {
                if (formattedName.charAt(formattedName.length() - 1) == 's') {
                    formattedName = formattedName.substring(0, formattedName.length() - 1);
                    for (UOM entry : uomsAbbrevDict.values()) {
                        if (entry.Name.toLowerCase().equals(formattedName)) {
                            uom = entry;
                            break;
                        }
                    }
                }
            }
        }
        return uom;
    }

    public static UOM getUOMFromUNCode(String name) {
        String formattedName = name.toUpperCase();
        UOM uom = uomsUNCodeDict.get(formattedName);
        if (uom == null) {
            for (UOM entry : uomsAbbrevDict.values()) {
                if (entry.UNCode.equalsIgnoreCase(formattedName)) {
                    uom = entry;
                    break;
                }
            }
        }
        return uom;
    }

    public static ArrayList<UOM> getList() {
        return new ArrayList<>(uomsAbbrevDict.values());
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/utility/uoms.kt
package opentraceability.utility

import opentraceability.OTLogger
import java.util.*
import org.json.*

object UOMS {
    private val uomsAbbrevDict = mutableMapOf<String, UOM>()
    private val uomsUNCodeDict = mutableMapOf<String, UOM>()

    init {
        load()
    }

    fun load() {
        try {
            // Load the subscriptions JSON
            val loader = EmbeddedResourceLoader()
            val jsonText = loader.readString("OpenTraceability", "OpenTraceability.Utility.Data.uoms.json")
            val jarr = JSONArray(jsonText)

            for (i in 0 until jarr.length()) {
                val juom = jarr.getJSONObject(i)
                val uom = UOM(juom)
                val lowerCaseAbbreviation = uom.Abbreviation.toLowerCase()
                val upperCaseUNCode = uom.UNCode.toUpperCase()

                if (!uomsAbbrevDict.containsKey(lowerCaseAbbreviation)) {
                    uomsAbbrevDict[lowerCaseAbbreviation] = uom
                } else {
                    println("Duplicate Unit abbreviation detected: ${uom.Abbreviation}")
                }

                if (!uomsUNCodeDict.containsKey(upperCaseUNCode)) {
                    uomsUNCodeDict[upperCaseUNCode] = uom
                } else {
                    println("Duplicate Unit UNCode detected: ${uom.UNCode}")
                }
            }
        } catch (ex: Exception) {
            opentraceability.OTLogger.error(ex)
            throw ex
        }
    }


    fun getBase(uom: UOM): UOM {
        return getBase(uom.UnitDimension)
    }

    fun getBase(dimension: String): UOM {
        for (entry in uomsAbbrevDict) {
            if (entry.value.UnitDimension == dimension && entry.value.isBase()) {
                return entry.value
            }
        }
        throw Exception("Failed to get base for dimension = $dimension")
    }

    fun getUOMFromName(name: String): UOM? {
        var uom: UOM? = null
        var formattedName = name.toLowerCase()
        if (formattedName == "count") {
            formattedName = "ea"
        }
        if (formattedName == "pound" || formattedName == "pounds" || formattedName == "ib") {
            formattedName = "lb"
        }
        if (formattedName[formattedName.length - 1] == '.') {
            formattedName = formattedName.substring(0, formattedName.length - 1)
        }
        if (uomsAbbrevDict.containsKey(formattedName)) {
            uom = uomsAbbrevDict[formattedName]
        } else {
            for (entry in uomsAbbrevDict) {
                if (entry.value.Name.toLowerCase() == formattedName) {
                    uom = entry.value
                    break
                }
            }
            if (uom == null) {
                if (formattedName[formattedName.length - 1] == 's') {
                    formattedName = formattedName.substring(0, formattedName.length - 1)
                    for (entry in uomsAbbrevDict) {
                        if (entry.value.Name.toLowerCase() == formattedName) {
                            uom = entry.value
                            break
                        }
                    }
                }
            }
        }
        return uom
    }

    fun getUOMFromUNCode(name: String): UOM? {
        val formattedName = name.toUpperCase()
        return uomsUNCodeDict[formattedName] ?: uomsAbbrevDict.values.find { it.UNCode.equals(formattedName, ignoreCase = true) }
    }

    val list: MutableList<UOM>
        get() = uomsAbbrevDict.values.toMutableList()
}
